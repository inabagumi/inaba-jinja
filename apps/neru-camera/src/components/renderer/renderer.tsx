import { makeStyles } from '@material-ui/core/styles'
import { Application, Sprite } from 'pixi.js'
import { Viewport } from 'pixi-viewport'
import React, {
  ReactElement,
  SyntheticEvent,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState
} from 'react'
import { Asset } from '../../context/asset-context'
import { ChromaKeyFilter } from '../../filters/ChromaKeyFilter'

const useStyles = makeStyles({
  container: {
    position: 'relative',
    width: '100%'
  },
  preloaded: {
    display: 'none'
  },
  preview: {
    display: 'block',
    height: 'auto',
    width: '100%'
  }
})

export type RefObject = {
  toBlob: (type: string, quality: number) => Promise<Blob>
}

export type Props = {
  asset: Asset
  cameraStream: MediaStream
}

const Renderer = forwardRef<RefObject, Props>(
  ({ asset, cameraStream }, ref): ReactElement => {
    const [viewport, setViewport] = useState<Viewport>()
    const classes = useStyles()

    const previewRef = useRef<HTMLCanvasElement>(null)
    const cameraRef = useRef<HTMLVideoElement>(null)
    const overlayRef = useRef<HTMLVideoElement>(null)

    const setup = useCallback(
      (event: SyntheticEvent<HTMLVideoElement, Event>) => {
        if (!previewRef.current) return

        const background = Sprite.from(event.currentTarget)

        const app = new Application({
          height: background.height,
          preserveDrawingBuffer: true,
          view: previewRef.current,
          width: background.width
        })
        const viewport = new Viewport({
          interaction: app.renderer.plugins.interaction,
          screenHeight: app.screen.height,
          screenWidth: app.screen.width,
          worldHeight: app.screen.height,
          worldWidth: app.screen.width
        })

        viewport
          .drag()
          .pinch()
          .wheel()
          .decelerate()

        app.stage.addChild(background)
        app.stage.addChild(viewport)

        setViewport(viewport)

        if (overlayRef.current) overlayRef.current.load()
      },
      []
    )

    const handleLoadedData = useCallback(
      (event: SyntheticEvent<HTMLVideoElement, Event>) => {
        if (!viewport) return

        const target = event.currentTarget
        const overlay = Sprite.from(target)
        const { keyColor } = target.dataset

        overlay.filters = [new ChromaKeyFilter(keyColor)]
        overlay.anchor.set(0.5)
        overlay.position.set(
          viewport.worldWidth * 0.5,
          viewport.worldHeight * 0.5
        )

        viewport.addChild(overlay)
      },
      [viewport]
    )

    useEffect(() => {
      if (!cameraRef.current) return

      cameraRef.current.srcObject = cameraStream
    }, [cameraStream])

    useImperativeHandle(ref, () => ({
      toBlob: (type = 'image/png', quality = 1.0): Promise<Blob> =>
        new Promise((resolve, reject): void => {
          if (!previewRef.current) {
            reject(new TypeError('Canvas is not mounted.'))
            return
          }

          const canvas = document.createElement('canvas')
          const context = canvas.getContext('2d')

          canvas.width = previewRef.current.width
          canvas.height = previewRef.current.height

          if (!context) {
            reject(new TypeError('Context does not exist.'))
            return
          }

          try {
            context.drawImage(previewRef.current, 0, 0)
            canvas.toBlob(
              blob => {
                if (!blob) {
                  reject(new TypeError('Blob is empty.'))
                  return
                }

                resolve(blob)
              },
              type,
              quality
            )
          } catch (error) {
            reject(error)
          }
        })
    }))

    return (
      <div className={classes.container}>
        <canvas className={classes.preview} ref={previewRef} />

        <video
          className={classes.preloaded}
          crossOrigin="anonymouse"
          muted
          onLoadedData={setup}
          playsInline
          ref={cameraRef}
        />
        <video
          className={classes.preloaded}
          crossOrigin="anonymouse"
          data-key-color={asset.keyColor}
          loop
          muted
          onLoadedData={handleLoadedData}
          playsInline
          preload="none"
          ref={overlayRef}
          src={asset.src}
        />
      </div>
    )
  }
)

export default Renderer
