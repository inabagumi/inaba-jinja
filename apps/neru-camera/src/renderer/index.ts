import {
  Color,
  LinearFilter,
  OrthographicCamera,
  PlaneGeometry,
  RGBFormat,
  ShaderMaterial,
  Scene,
  WebGLRenderer,
  Mesh,
  MeshBasicMaterial,
  VideoTexture,
} from 'three';

const fragmentShader = `precision mediump float;
uniform sampler2D texture;
uniform vec3 keyColor;
varying vec2 vUv;
void main(void) {
  vec4 smpColor = texture2D(texture, vUv);
  float diff = length(keyColor - smpColor.rgb);
  if (diff < 0.6) {
    discard;
  } else {
    gl_FragColor = smpColor;
  }
}
`;
const vertexShader = `precision mediump float;
varying vec2 vUv;
void main(void) {
  vUv = uv;
  gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
}
`;

export default class Renderer {

  private background: Mesh;
  private camera: OrthographicCamera;
  private overlay: Mesh;
  private renderer: WebGLRenderer;
  private sceen: Scene;

  constructor(canvas: HTMLCanvasElement) {
    this.renderer = new WebGLRenderer({ canvas, preserveDrawingBuffer: true });
    this.sceen = new Scene();

    const { height, width } = this.renderer.getSize();
    this.setCamera(width, height);
  }

  render() {
    this.renderer.render(this.sceen, this.camera);
  }

  setCamera(width: number, height: number) {
    this.camera = new OrthographicCamera(
      width / -2,
      width / 2,
      height / 2,
      height / -2,
      1,
      2000,
    );

    this.camera.position.z = 500;
  }

  setBackground(media: HTMLVideoElement) {
    const { videoHeight, videoWidth } = media;

    this.renderer.setSize(videoWidth, videoHeight, false);
    this.setCamera(videoWidth, videoHeight);

    const texture = new VideoTexture(media);
    texture.minFilter = LinearFilter;
    texture.magFilter = LinearFilter;
    texture.format = RGBFormat;

    const geometry = new PlaneGeometry(videoWidth, videoHeight, 1, 1);
    const material = new MeshBasicMaterial({ map: texture });
    this.background = new Mesh(geometry, material);

    this.sceen.add(this.background);
  }

  setOverlay(media: HTMLVideoElement) {
    const { videoHeight, videoWidth } = media;

    const texture = new VideoTexture(media);
    texture.minFilter = LinearFilter;
    texture.magFilter = LinearFilter;
    texture.format = RGBFormat;

    const geometry = new PlaneGeometry(videoWidth, videoHeight, 1, 1);
    const material = new ShaderMaterial({
      fragmentShader,
      vertexShader,
      transparent: true,
      uniforms: {
        keyColor: {
          type: 'c',
          value: new Color(0x00ff00),
        },
        texture: {
          type: 't',
          value: texture,
        },
      },
    });
    this.overlay = new Mesh(geometry, material);
    this.overlay.scale.x = this.overlay.scale.y = 0.5;

    this.sceen.add(this.overlay);
  }

  start() {
    this.renderer.setAnimationLoop(() => this.render());
  }

  updateTransform(scale: number, x: number, y: number) {
    const { clientHeight, clientWidth, height, width } = this.renderer.domElement;
    const deviation = 100 * ((scale / 0.5) - 1);

    this.overlay.position.x = (x + deviation) / (clientWidth / width);
    this.overlay.position.y = -((y + deviation) / (clientHeight / height));
    this.overlay.scale.x = this.overlay.scale.y = scale;
  }
}
