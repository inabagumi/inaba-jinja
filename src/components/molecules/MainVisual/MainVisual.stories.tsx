import { storiesOf } from '@storybook/react'
import React from 'react'
import MainVisual from './MainVisual'

storiesOf('MainVisual', module).add('default', () => <MainVisual />)
