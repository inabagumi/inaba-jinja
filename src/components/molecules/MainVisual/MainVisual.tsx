import styled from '@emotion/styled'
import React, { FC } from 'react'
import { Helmet } from 'react-helmet'
import mainVisual from '../../../assets/main-visual.jpg'
import webpMainVisual from '../../../assets/main-visual.webp'
import lqipMainVisual from '../../../assets/main-visual@lqip.jpg'
import Container from '../../atoms/Container'

const Content = styled(Container)`
  align-items: center;
  bottom: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  left: 0;
  padding: 0 1rem;
  position: absolute;
  right: 0;
  top: 0;
`

const Cover = styled('div')`
  background-color: transparent;
  background-image: url("${lqipMainVisual}");
  background-position: center;
  background-repeat: no-repeat;
  background-size: cover;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;

  &::after {
    background-color: var(--hero-overlay-color);
    bottom: 0;
    content: '';
    display: block;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  }
`

const CoverImage = styled('picture')`
  img {
    bottom: 0;
    display: block;
    height: 100%;
    left: 0;
    object-fit: cover;
    position: absolute;
    right: 0;
    top: 0;
    width: 100%;
  }
`

const Root = styled('div')`
  --hero-overlay-color: rgba(0, 0, 0, 0.54);

  background-color: var(--hero-overlay-color);
  color: #fff;
  min-height: 80vh;
  position: relative;

  @media (min-width: 960px) {
    & {
      min-height: 100vh;
    }
  }
`

const Hero: FC = ({ children }) => {
  return (
    <>
      <Helmet>
        <link
          as="image"
          href={webpMainVisual}
          rel="preload"
          type="image/webp"
        />
      </Helmet>

      <Root>
        <Cover>
          <CoverImage>
            <source srcSet={webpMainVisual} type="image/webp" />

            <img
              alt=""
              className="cover__image"
              height="600"
              role="presentation"
              src={mainVisual}
              width="800"
            />
          </CoverImage>
        </Cover>

        <Content>{children}</Content>
      </Root>
    </>
  )
}

export default Hero
