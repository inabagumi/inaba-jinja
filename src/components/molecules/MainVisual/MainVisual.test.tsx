import React, { FC } from 'react'
import { IntlProvider } from 'react-intl'
import renderer from 'react-test-renderer'
import MainVisual from './MainVisual'

const App: FC = ({ children }) => (
  <IntlProvider locale="en">{children}</IntlProvider>
)

describe('MainVisual', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <App>
          <MainVisual />
        </App>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
