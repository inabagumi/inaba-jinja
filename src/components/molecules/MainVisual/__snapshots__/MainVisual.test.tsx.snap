// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainVisual renders correctly 1`] = `
<div
  className="jsx-4256674884 hero"
>
  <div
    className="jsx-4256674884 cover"
  >
    <picture
      className="jsx-4256674884"
    >
      <source
        className="jsx-4256674884"
        srcSet="test-file-stub"
        type="image/webp"
      />
      <img
        alt=""
        className="jsx-4256674884 cover__image"
        height="600"
        role="presentation"
        src="test-file-stub"
        width="800"
      />
    </picture>
  </div>
  <div
    className="jsx-4256674884 hero__content"
  />
</div>
`;
